### ISSUE TITLE
ITS NOT A BUG cant import LogisticREGRESSION TRAINER OTHER PACKage.?

### ISSUE THREAD

**Issue by nikkisingh111333** [oracle/tribuo#54]
Its not a bug i know.but i m unable to import all the packages after successfully dowloaded all dependency from maven.logisticregression is not present in classpath and other evaluation methods are not impoting dont know why..some are already imported but others are not why?... preprocessing stuffs like Grid SearchCv is not present

**Comment by Craigacp** [comment]
What's the classpath that you're using (or alternatively what maven dependency are you using)? What's the error message you get, is at runtime or out of Maven/Gradle/javac at compile time?

Tribuo doesn't have an equivalent to `GridSearchCV`, we have a `CrossValidation` class in `org.tribuo.evaluation` in the `tribuo-core` artifact. We're considering what the right API is for hyperparameter optimization, but something like grid search is pretty easy for users to implement themselves on top of Tribuo. Tribuo has a selection of feature preprocessors in the `org.tribuo.transform` and you can perform some transformations in a `DataSource` depending on which one you use to load the data in (e.g. `RowProcessor` can perform many different extraction and transformation operations as it loads in columnar data, and the `TextPipeline` can transform textual inputs before featurising them.

**Comment by nikkisingh111333** [comment]
hey thanks for support,i m using tribuo-all dependency and following your
tutorial on of your website.doing exact same thing as described in the
tutorial.im not getting error but some dependencies packages are not
resolving like models logisticregrwssion and labelfactory others are
working well...plz sort me out.i m new in ML world.one more thing can your
version of cross validation work exact like gridsearch.like finding good
trainer and regulazed param for model etc.

**Comment by Craigacp** [comment]
Are they failing to resolve at compile time or runtime, or are you running the jupyter notebook itself and getting an error?

Tribuo's `CrossValidation` class performs a similar job to scikit-learn's [cross_validate](https://scikit-learn.org/stable/modules/cross_validation.html#the-cross-validate-function-and-multiple-metric-evaluation) function, and is designed for model evaluation to get an idea of the variance of the evaluation metrics. If we add a grid search hyperparameter optimizer then it will be a different class and may or may not use cross validation as the model evaluation component.

**Comment by nikkisingh111333** [comment]
Heres i ve share two screenshot..in one pic i have copied your web site
code ..RED underlined constructor indicates cannot resolve error.some are
resolving but some are not.
Another screenshot is my pom.xml with the tribou dependency and weka
dependency(which i was using earliar but feels like an old ML library with
no proper documentations).

**Comment by Craigacp** [comment]
You need to attach the screenshots to the issue on github, they didn't come through when you attached them to the email replying to my comment.

**Comment by nikkisingh111333** [comment]
Okk heres my screenshots..

**Comment by nikkisingh111333** [comment]
Okk heres my screenshots..
![Screenshot (34)](https://user-images.githubusercontent.com/49682940/95024460-bd931880-06a0-11eb-9309-f927f0da78ff.png)
![Screenshot (35)](https://user-images.githubusercontent.com/49682940/95024464-c08e0900-06a0-11eb-8bff-3098454a3914.png)

In 1st one RED underlined indicates no type resolve error..while others are resolving but some are not.
2nd one is my pom.xml with tribuo ALL DEPENDENCY AND weka.
why this is happenin ive also tried to rebuilt my project but not resolving ..its showing SGD linear trainer resolved but No Logistic Or tree..
..What's the correct way to kick start my tribuo journey and what other stuffs i will need ..sort me out please.

**Comment by Craigacp** [comment]
What imports do you have in the `Tribuo.ML.java` file? The ones from the [irises tutorial](https://tribuo.org/learn/4.0/tutorials/irises-tribuo-v4.html) should be sufficient. Your IDE might not be auto-importing `LabelFactory` etc if there are multiple classes with those names on your classpath. In particular `org.tribuo.classification.Label` usually does not auto-import because it collides with `java.awt.Label`.

I'd recommend you work through the tutorial notebooks - https://github.com/oracle/tribuo/tree/v4.0.X-release-branch/tutorials, as those are all running examples. Note that on Windows the XGBoostTrainer used in the regression and configuration tutorials is not available as the XGBoost library on Maven Central doesn't provide a Windows binary (though you can compile it by hand and add it to the classpath).

**Comment by nikkisingh111333** [comment]
Sorry for the bad news but.....i can't see any org.tribuo.classification.Label
*CLASS OR package in my classpath i have tribou.core jar package and
tribou.data jar tribuo-json jar  and other tribuo jars but nothing like
org.tribuo.classification in any core jar...why its happening i ve tried to
refresh every dependency but not resolving. *

**Comment by Craigacp** [comment]
The `tribuo-all` dependency pulls in `tribuo-classification-core` which is where `LabelFactory`, `Label`, `LabelEvaluator` and `LabelEvaluation` live, you can see the pom file for it here - https://search.maven.org/artifact/org.tribuo/tribuo-all/4.0.1/pom. I'm not sure why it's not loaded for you, try checking to see if maven has already downloaded the `tribuo-classification-core` jar to your local cache. `LogisticRegressionTrainer` lives in `tribuo-classification-sgd` which is also a dependency of `tribuo-all` so look for that jar too.

**Comment by nikkisingh111333** [comment]
I have manually added dependency Again some dependency resolved on top
successfully while RED underlined are not and Classification.core is not
resolving again..same error
HERES A QUICK SCREENSHOT.

**Comment by Craigacp** [comment]
The screenshot isn't attached, you need to add it via the GitHub web interface.

**Comment by nikkisingh111333** [comment]
okkk heres the screenshot...again error is happening NO CLASSIFICATION DEPENDENCY IS RESOLVING (indicated by RED undelines) BUT OTHERS ARE RESOLVING GOOD.
![Screenshot (36)](https://user-images.githubusercontent.com/49682940/95178358-f932fd00-07dc-11eb-9d88-5388172fa684.png)

**Comment by nikkisingh111333** [comment]
I HAVE SENT IT TO YOU VIA GITHUB CHECK IT OUT AND..PLEASE HELP ME OUT.I VE
TRIED EVERYTHING

**Comment by Craigacp** [comment]
The jar files are available on Maven Central e.g. https://search.maven.org/artifact/org.tribuo/tribuo-classification-core/4.0.1/jar, or you can compile it from source by checking out the repo and the v4.0.1 tag and running `mvn install -DskipTests`. If I had to guess, then I think you've got half downloaded files in your maven cache and it's failing to redownload them for some reason. Try deleting the `.m2/repository/org/tribuo` directory from your user directory and re-running maven (or make your IDE reimport your pom file).

**Comment by nikkisingh111333** [comment]
YES .it worked i can see classification pack now and its resolving after
deleting .m2/repo/tribuo...but still some dependency are missing...this
time this one is getting dirty
.."com.oracle.labs.mlrg.olcut.provenance.Provenance cannot be
resolved."...btw thanks for your support ..i was searching for 2 weeks
about ml with java and now with this i can feel like i dont need python
....rapidminer with tribuo best duo i ve got..i hope you will update
documentations regularly.cause java without a proper docs is a
nightmare...one more thing is there any lovely tool for visualization in
java like mltplot OR seaborn.I found one UJMP but didnt like their
docs..any guess for me?.

**Comment by Craigacp** [comment]
The provenance code is in `com/oracle/labs/olcut/` so if that also didn't download correctly then delete that folder in your maven cache and try again.

We're working on the documentation, and people are welcome to contribute docs fixes too. The tribuo website is also hosted on GitHub - https://github.com/oracle/tribuo-site.

We've not figured out a good plotting solution for Tribuo yet. I've looked at XChart and jzy3d in the past but wasn't particularly happy with the docs for either of them. There is basic plotting support built into JavaFX, but we've not had chance to investigate it yet.

**Comment by nikkisingh111333** [comment]
okk so is there no chance i can use plotting in java.so.i had to stick with
gui tools like rapidminer?

**Comment by Craigacp** [comment]
No, you can definitely plot things in Java, it's just not straightforward like it is in Python with matplotlib and numpy because of the lack of a commonly accepted ndarray type in Java. That means things need a bit of glue between the plotting library and whatever the output is that you're trying to plot. We've not written that glue for Tribuo yet, because we've not figured out which plotting library is appropriate to depend on.

**Comment by nikkisingh111333** [comment]
okk i got it thanks mr.adam i will look for other alternative in plotting
(RapidMiner is good since i only need preprocesssing and roc auc for now)
currently i m getting my head in tribuo config.
can you  please explain me tribou config is kind a grid search or
docker-compose like stuff i guess where we put all our stuff in 1 file and
run.is this what it does?

**Comment by Craigacp** [comment]
Tribuo's configuration system underlies the command line options and provenance systems, and allows you to specify the parameters of an experiment (e.g. the data source, any transformations applied to the dataset, training hyperparameters etc). There are examples of how to use it in the various small [helper programs](https://github.com/oracle/tribuo/blob/main/docs/HelperPrograms.md) that come with Tribuo, plus the [configuration tutorial](https://github.com/oracle/tribuo/blob/main/tutorials/configuration-tribuo-v4.ipynb). You could certainly use the configuration system to build a grid search, and we have done so in the past (though that code depends on various outdated clustering technologies so it isn't going to be released). This thread seems to have strayed from the initial dependency issue, so I'm going to close it now. Open a new issue if you have other problems with using Tribuo.

**Comment by nikkisingh111333** [comment]
okk thanks mr. adam i will open another issue whenever i will need you
thanks for your great support.😊
